### ISSUE TITLE
Invalidate and push should return a value

### ISSUE THREAD

**Issue by nikolavp** [ben-manes/caffeine#83]
Currently invalidate doesn't return the value and put does not return the previous value in the cache. We want to move from ConcurrentLinkedHashMap but the new API is making some things really hard. Is there a reason why those two methods are void at the moment.

If they are void by design, I think that this should be clearly stated in the documentation.

**Comment by ben-manes** [comment]
The `Cache` interface is more philosophically pure. A cache holds transient, recomputable state and an ideal usage wouldn't need to inspect it directly. Like a CPU cache, data would flow in and out to transparently improve performance. If the value was non-local (remote, persisted) then returning it may be costly, especially if seldomly used.

Most real code quickly gets dirty, so only providing a pure interface would be frustrating by blocking you from getting your job done. `Cache.asMap()` is an escape hatch which should provide you with everything you need. This also keeps the conceptual weight low because it builds on known interfaces and the new ones you'd learn are all small.

Its understandably subtle and if you have any recommendations for how to improve the documentation to make it more obvious I'd appreciate it.

**Comment by drcrallen** [comment]
@ben-manes I found this [video from the JCache guys](https://blogs.oracle.com/theaquarium/entry/jcache_is_final_i_repeat) to be pretty good

**Comment by ben-manes** [comment]
The JCache 1.x API is a bit weird. Their [original design](http://www.javadoc.io/doc/net.sf.jsr107cache/jsr107cache/1.1) extended `Map`, but that is confusing because it breaks the contract (`get` is not passive so equality issues). I described a few of these quirks in these [slides](http://concurrentlinkedhashmap.googlecode.com/files/ConcurrentCachingAtGoogle.pdf) for why we introduced the `Cache` interface in Guava.

The final JCache APIs avoid map and provide different methods that return `void` or `V`. They try to bypass the Map issues, except allow full traversal by being `Iterable`. That's weird given some caches (like memcached) don't allow this for obvious reasons. It also means a full Map interface can be emulated, but slower (e.g. there is no size so its O(n)). Its odd to both reject the Java Collection Framework while still providing all of the controversial methods so that it was reinvented.

Instead we tried to design an API to feel like a natural extension to Java's collections. I think `cache.asMap()` is a nice compromise between the purist and pragmatist views. It promotes the preferred style for common (simple) cases, while solving @nikolavp issue by using `cache.asMap().put(k, v)` and `cache.asMap().remove(k)`. It keeps the interfaces short and simple, but it is easy to miss that the power of `Map` is a hop away.
